### PR TITLE
Feature: Adding Bucket Name Parameter to Storage Class to Support Bucket Name of Your Choice

### PR DESCRIPTION
**Pull request by dmakeroam**
I have added bucket name parameter to support config bucket name in storage class to choose bucket name of your choice while mouting

### CODE REVIEW

**Comment by dmakeroam** [comment]
@majst01 Please help to review this

File: pkg/s3/controllerserver.go [hunk 1]
```diff
@@ -41,6 +41,8 @@ func (cs *controllerServer) ControllerGetVolume(ctx context.Context, req *csi.Co
 }
 
 func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+	s3, _ := newS3ClientFromSecrets(req.GetSecrets())
+	bucketName := *s3.bucketName
```
Comment (majst01) [inline pkg/s3/controllerserver.go:45]: nilcheck

File: pkg/s3/controllerserver.go [hunk 2]
```diff
@@ -49,8 +51,12 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	}
 
 	// Check arguments
-	if len(volumeID) == 0 {
+	if len(bucketName) != 0 {
```
Comment (majst01) [inline pkg/s3/controllerserver.go:54]: if bucketName != ""

File: pkg/s3/controllerserver.go [hunk 3]
```diff
@@ -49,8 +51,12 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+		volumeID = bucketName
+	} else if len(volumeID) == 0 {
```
Comment (majst01) [inline pkg/s3/controllerserver.go:56]: no else if please

File: pkg/s3/controllerserver.go [hunk 4]
```diff
@@ -75,11 +81,17 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 }
 
 func (cs *controllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	s3, err := newS3ClientFromSecrets(req.GetSecrets())
+	bucketName := *s3.bucketName
```
Comment (majst01) [inline pkg/s3/controllerserver.go:85]: nilcheck

File: pkg/s3/controllerserver.go [hunk 5]
```diff
@@ -75,11 +81,17 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
 	volumeID := req.GetVolumeId()
 
 	// Check arguments
-	if len(volumeID) == 0 {
+	if len(bucketName) != 0 {
```
Comment (majst01) [inline pkg/s3/controllerserver.go:89]: if bucketName !=""

File: pkg/s3/controllerserver.go [hunk 6]
```diff
@@ -75,11 +81,17 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+		volumeID = bucketName
+	} else if len(volumeID) == 0 {
```
Comment (majst01) [inline pkg/s3/controllerserver.go:91]: no else if

File: pkg/s3/nodeserver.go [hunk 7]
```diff
@@ -86,9 +86,15 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 		targetPath, deviceID, readOnly, volumeID, attrib, mountFlags)
 
 	s3, err := newS3ClientFromSecrets(req.GetSecrets())
+	bucketName := *s3.bucketName
```
Comment (majst01) [inline pkg/s3/nodeserver.go:89]: nilcheck

File: pkg/s3/nodeserver.go [hunk 8]
```diff
@@ -86,9 +86,15 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
 	if err != nil {
 		return nil, fmt.Errorf("failed to initialize S3 client: %w", err)
 	}
+	if len(bucketName) != 0 {
```
Comment (majst01) [inline pkg/s3/nodeserver.go:93]: if bucketName != ""

File: pkg/s3/nodeserver.go [hunk 9]
```diff
@@ -86,9 +86,15 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
+		volumeID = bucketName
+	} else {
+		return nil, status.Error(codes.InvalidArgument, "Bucket Name missing in request")
```
Comment (majst01) [inline pkg/s3/nodeserver.go:96]: this is wrong

File: pkg/s3/nodeserver.go [hunk 10]
```diff
@@ -150,9 +156,15 @@ func (ns *nodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 		return nil, status.Error(codes.Internal, fmt.Sprintf("unable to create mkdir directory for %q err:%v", stagingTargetPath, err))
 	}
 	s3, err := newS3ClientFromSecrets(req.GetSecrets())
+	bucketName := *s3.bucketName
```
Comment (majst01) [inline pkg/s3/nodeserver.go:159]: nilcheck

**Review by majst01 — CHANGES_REQUESTED** [review]

File: pkg/s3/nodeserver.go (hunk 9 above)
Comment (dmakeroam) [reply pkg/s3/nodeserver.go:99]: Why this is wrong?

**Comment by dmakeroam** [comment]
@majst01 I fixed my code, please help to review

File: pkg/s3/nodeserver.go (hunk 9 above)
Comment (dmakeroam) [reply pkg/s3/nodeserver.go:99]: I tested my code on my side for bucket name mounting and it worked well

File: pkg/s3/controllerserver.go [hunk 11]
```diff
@@ -41,15 +41,30 @@ func (cs *controllerServer) ControllerGetVolume(ctx context.Context, req *csi.Co
 }
 
 func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
-	volumeID := req.GetName()
+	s3, err := newS3ClientFromSecrets(req.GetSecrets())
+	bucketName := ""
+	if s3 != nil {
+		bucketName = *s3.bucketName
```
Comment (majst01) [inline pkg/s3/controllerserver.go:47]: this panics if bucketName is nil

File: pkg/s3/controllerserver.go [hunk 12]
```diff
@@ -41,15 +41,30 @@ func (cs *controllerServer) ControllerGetVolume(ctx context.Context, req *csi.Co
...
+	}
+	if err != nil {
```
Comment (majst01) [inline pkg/s3/controllerserver.go:49]: errcheck should be nearby the err

File: pkg/s3/controllerserver.go [hunk 13]
```diff
@@ -41,15 +41,30 @@ func (cs *controllerServer) ControllerGetVolume(ctx context.Context, req *csi.Co
...
 	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME); err != nil {
 		klog.Infof("invalid create volume req: %v", req)
 		return nil, err
 	}
 
 	// Check arguments
+	if bucketName != "" {
+		volumeID = bucketName
+	} else {
+		return nil, status.Error(codes.InvalidArgument, "Bucket Name missing in request")
```
Comment (majst01) [inline pkg/s3/controllerserver.go:63]: error messages should not start with uppercase

File: pkg/s3/controllerserver.go [hunk 14]
```diff
@@ -75,10 +90,25 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 }
 
 func (cs *controllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
-	volumeID := req.GetVolumeId()
+	s3, err := newS3ClientFromSecrets(req.GetSecrets())
+	bucketName := ""
+	if s3 != nil {
+		bucketName = *s3.bucketName
```
Comment (majst01) [inline pkg/s3/controllerserver.go:96]: see above

File: pkg/s3/controllerserver.go [hunk 15]
```diff
@@ -75,10 +90,25 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
...
+	}
+	if err != nil {
```
Comment (majst01) [inline pkg/s3/controllerserver.go:98]: see above

**Review by majst01 — CHANGES_REQUESTED** [review]

File: pkg/s3/controllerserver.go (hunk 13 above)
Comment (dmakeroam) [reply pkg/s3/controllerserver.go]: Done

File: pkg/s3/controllerserver.go (hunk 15 above)
Comment (dmakeroam) [reply pkg/s3/controllerserver.go:96]: Done

File: pkg/s3/controllerserver.go (hunk 14 above)
Comment (dmakeroam) [reply pkg/s3/controllerserver.go:94]: Done

**Comment by dmakeroam** [comment]
Done fixed as your reviewed @majst01

**Comment by majst01** [comment]
After thinking again, this PR makes not much sense, because with this only one PVC per storageclass is possible. 
What is you exact requirement ?

**Comment by dmakeroam** [comment]
> After thinking again, this PR makes not much sense, because with this only one PVC per storageclass is possible. What is you exact requirement ?

I would like to mount a S3 to my pod with specific S3 bucket name to align with my customer compliance, it would not be good to name S3 bucket as PVC name for my use case.

**Comment by majst01** [comment]
> > After thinking again, this PR makes not much sense, because with this only one PVC per storageclass is possible. What is you exact requirement ?
> 
> @majst01 I would like to mount a S3 to my pod with specific S3 bucket name to align with my customer compliance, it would not be good to name S3 bucket as PVC name for my use case.

But what happens then if you  create a second PVC ??
So with this the general usability is basically gone

**Comment by dmakeroam** [comment]
> > > After thinking again, this PR makes not much sense, because with this only one PVC per storageclass is possible. What is you exact requirement ?
> > 
> > 
> > @majst01 I would like to mount a S3 to my pod with specific S3 bucket name to align with my customer compliance, it would not be good to name S3 bucket as PVC name for my use case.
> 
> But what happens then if you create a second PVC ?? So with this the general usability is basically gone

For my use case is generally specific, I have a use case that one pod will use one pvc for mounting data in the blob storage like S3 (This app has only replica). Another use case is I use statefulset with blob mounting to get JKS files for Kafka

**Comment by dmakeroam** [comment]
From your mention "because with this only one PVC per storageclass is possible."

For my Kafka use case, that is statefulset, as the statefulset nature, it will create per pod pvc that is mounting to the same blob using storage class. I think this is possibly possible, because we don't share the pvc

**Comment by majst01** [comment]
> From your mention "because with this only one PVC per storageclass is possible."
> 
> @majst01 For my Kafka use case, that is statefulset, as the statefulset nature, it will create per pod pvc that is mounting to the same blob using storage class. I think this is possibly possible, because we don't share the pvc and we just mouting the JKS files (not modifying the blob contents)

Sure, this is the contract of a sts. But the contract of csi is that configuration parameters of a storageclass apply to all PVCs created and with your PR you will never be able to create more than one PVC in the whole cluster with this storageclass. Therefore this PR will not be merged.

**Comment by dmakeroam** [comment]
> > From your mention "because with this only one PVC per storageclass is possible."
> > @majst01 For my Kafka use case, that is statefulset, as the statefulset nature, it will create per pod pvc that is mounting to the same blob using storage class. I think this is possibly possible, because we don't share the pvc and we just mouting the JKS files (not modifying the blob contents)
> 
> Sure, this is the contract of a sts. But the contract of csi is that configuration parameters of a storageclass apply to all PVCs created and with your PR you will never be able to create more than one PVC in the whole cluster with this storageclass. Therefore this PR will not be merged.

@majst01 How can we change the contract to support multiple PVCs?

**Pull request closed without merging**